experimental/components/VideoPlayer/library/src/desktopMain/kotlin/org/jetbrains/compose/videoplayer/DesktopVideoPlayer.kt

```diff
@@ -1,63 +1,137 @@
+    LaunchedEffect(isFullscreen) {
+        if (mediaPlayer is EmbeddedMediaPlayer) {
+            /*
+             * To be able to access window in the commented code below,
+             * extend the player composable function from WindowScope.
+             * See https://github.com/JetBrains/compose-jb/issues/176#issuecomment-812514936
+             * and its subsequent comments.
+             *
+             * We could also just fullscreen the whole window:
+             * `window.placement = WindowPlacement.Fullscreen`
+             * See https://github.com/JetBrains/compose-multiplatform/issues/1489
+             */
+            // mediaPlayer.fullScreen().strategy(ExclusiveModeFullScreenStrategy(window))
+            mediaPlayer.fullScreen().toggle()
         }
     }
-    SideEffect {
-        mediaPlayerComponent.mediaPlayer().media().play(url)
+    DisposableEffect(Unit) { onDispose { mediaPlayer.release() } }
+    SwingPanel(
+        factory = factory,
+        background = Color.Transparent,
+        modifier = modifier
+    )
+    return mediaPlayer.produceProgress()
+}
+
+private fun Float.toPercentage() = (this * 100).roundToInt()
+
+/**
+ * See https://github.com/caprica/vlcj/issues/887#issuecomment-503288294
+ * for why we're using CallbackMediaPlayerComponent for macOS.
+ */
+@Composable
+private fun initializeMediaPlayerComponent(): Component = remember {
+    NativeDiscovery().discover()
+    if (isMacOS()) {
+        CallbackMediaPlayerComponent()
+    } else {
+        EmbeddedMediaPlayerComponent()
     }
+}
 
-    DisposableEffect(Unit) {
-        onDispose {
-            mediaPlayerComponent.mediaPlayer().release()
+/**
+ * We play the video on finish (so the player is kind of idempotent),
+ * unless the [onFinish] callback stops the playback.
+ * Using `mediaPlayer.controls().repeat = true` did not work as expected.
+ */
+@Composable
+private fun MediaPlayer.setupVideoFinishHandler(onFinish: (() -> Unit)?) =
+    DisposableEffect(onFinish) {
+        val listener = object : MediaPlayerEventAdapter() {
+            override fun stopped(mediaPlayer: MediaPlayer) {
+                onFinish?.invoke()
+                mediaPlayer.controls().play()
+            }
         }
+        events().addMediaPlayerEventListener(listener)
+        onDispose { events().removeMediaPlayerEventListener(listener) }
     }
 
-    return SwingPanel(
-        background = Color.Transparent,
-        modifier = Modifier.fillMaxSize(),
-        factory = {
-            mediaPlayerComponent
+/**
+ * Checks for and emits video progress every 50 milliseconds.
+ * Note that it seems vlcj updates the progress only every 250 milliseconds or so.
+ *
+ * Instead of using `Unit` as the `key1` for [produceState],
+ * we could use `media().info()?.mrl()` if it's needed to re-launch
+ * the producer (for whatever reason) when the url (aka video) changes.
+ *
+ */
+@Composable
+private fun MediaPlayer.produceProgress() =
```

Comment:
Could you please add the return type explictly?

Reply:
Sure.